cpplint.py

```diff
@@ -996,19 +1025,41 @@ def ParseNolintSuppressions(filename, raw_line, linenum, error):
     linenum: int, the number of the current line.
     error: function, an error handler.
   """
-  matched = Search(r'\bNOLINT(NEXTLINE)?\b(\([^)]+\))?', raw_line)
+  matched = Search(r'\bNOLINT(NEXTLINE|BEGIN|END)?\b(\([^)]+\))?', raw_line)
   if matched:
-    if matched.group(1):
-      suppressed_line = linenum + 1
+    no_lint_type = matched.group(1)
+    if no_lint_type == 'NEXTLINE':
+      def process_category(category):
+        _error_suppressions.setdefault(category, set()).add(linenum + 1)
+    elif no_lint_type == 'BEGIN':
+      if _block_error_suppressions:
+        error(filename, linenum, 'readability/nolint', 5,
+              'NONLINT block already defined on line %s' % _block_error_suppressions.linenum)
+      else:
+        _block_error_suppressions.linenum = linenum
+
+      def process_category(category):
+        _block_error_suppressions.add(category)
+    elif no_lint_type == 'END':
+      if not _block_error_suppressions:
+        error(filename, linenum, 'readability/nolint', 5, 'Not in a NOLINT block')
+
+      def process_category(category):
+        if category is not None:
```

Comment:
technically we do not need to check if anything is set, we could just ignore... So if some user wishes to add the categories there for documentation purposes, we don't make it weird. Checking this here would mainly help us introduce the differentiated end with nested blocks later if we introduced nested blocks.

Did you have another reason to check and error here?  I am not strongly opposed right now, need to think about it more.

Reply:
Yes I wanted to make it explicit that categories were not honored in the end statement and to prevent possible misleading no lint blocks like the one below where the END actually ends both cat1 and cat2. If END blocks for categories were added that would change the behavior of a block like below
```
...
//NOLINTBEGIN(cat1,cat2)
...
//NOLINTEND(cat1)
...
```